Automattic/jetpack
# Beta plugin: set the JETPACK_AUTOLOAD_DEV constant when changing Jetpack versions

Fixes #18407

#### Changes proposed in this Pull Request:
The Jetpack autoloader prefers to load stable versions of packages over development versions. This behavior can be overridden by setting the `JETPACK_AUTOLOAD_DEV` constant to true.

When testing development versions of Jetpack using the Beta plugin, we need to make sure that the `JETPACK_AUTOLOAD_DEV` constant is set to true so the development versions of packages will be used. Let's make the Jetpack Beta plugin do this for us:

 * Add a new option, `jetpack_autoload_dev`, which is set when a Jetpack version is activated.
    * Set the option to false when the stable or release candidate versions are activated. These Jetpack versions use stable versions of the packages.
    * Set the option to true when development versions of Jetpack are activated.

 * Set the `JETPACK_AUTOLOAD_DEV` constant using the value of the `jetpack_autoload_dev` constant when
   the Jetpack Beta plugin is loaded.

#### Jetpack product discussion
* n/a

#### Does this pull request change what data or activity we track or use?
* no

#### Testing instructions:
#### Reproduce a fatal caused by stable versions being used over dev versions.
1. Install and activate the stable version of Jetpack Beta. The bleeding edge version of Jetpack should be activated.
2. We're going to cause a fatal, so have SSH/FTP access ready so you can deactivate plugins/delete files.
3. Install and activate the latest version of Woo. See the fatal.
4. Deactivate or delete Woo.

#### Test this PR
1. Install and activate this branch of Jetpack Beta.
2. Install and activate the latest version of Woo.
3. Test different versions of Jetpack using Jetpack Beta: stable ([don't update stable](https://github.com/Automattic/jetpack/issues/18496) if you're testing in Jetpack's Docker development environment), release candidate, bleeding edge, and a development version. Verify that no fatal errors are generated.
4. Delete the `jetpack_autoload_dev` option before switching Jetpack versions to verify that no fatals are generated before the option is set.
     `wp option delete jetpack_autoload_dev`

#### Proposed changelog entry for your changes:
* tbd
